Introduce Efficient Buffer Writer with ReadOnlySequence<byte> Access

paulomorgado opened the pull request:
### 🛠️ Introduce Efficient Buffer Writer with ReadOnlySequence Access

#### 📌 Summary

This PR introduces a new class that implements `IBufferWriter<byte>` and provides a `GetReadOnlySequence()` method to expose the written data as a `ReadOnlySequence<byte>`. This replaces the previous approach of using `IEnumerable<ReadOnlyMemory<byte>>` for representing sequences of byte buffers.

#### ✅ Why This Change?

The new design offers **significant performance and usability improvements** over the previous `IEnumerable<ReadOnlyMemory<byte>>` approach:

---

### 🚀 Benefits of `IBufferWriter<byte>` + `ReadOnlySequence<byte>`

1. **Zero-Copy Efficiency**  
   As the current implementation using `IEnumerable<ReadOnlyMemory<byte>>`, `ReadOnlySequence<byte>` is designed for high-performance scenarios. It allows consumers to traverse a sequence of memory segments without copying data, which is ideal for parsers, serializers, and network protocols.

2. **Better Interoperability**  
   `ReadOnlySequence<byte>` is the standard input for many high-performance .NET APIs (e.g., `System.IO.Pipelines`, `System.Text.Json`, `System.Buffers`). This makes the new class more compatible with modern .NET libraries.

3. **Fine-Grained Control Over Memory**  
   `IBufferWriter<byte>` allows producers to write directly into pre-allocated buffers, reducing allocations and enabling better memory reuse.

4. **Improved Performance Characteristics**  
   - Avoids the overhead of enumerating `IEnumerable<ReadOnlyMemory<byte>>`.
   - Enables contiguous or segmented memory layouts with minimal overhead.
   - Supports pooling and span-based writing patterns.

5. **Cleaner API Surface**  
   The combination of `IBufferWriter<byte>` and `ReadOnlySequence<byte>` clearly separates the concerns of writing and reading, leading to more maintainable and testable code.

belveder79 committed:
Enhance memory management and performance across the codebase. Transitioned from MemoryStream to PooledBufferWriter in JpegPayloadTests.cs for better memory handling. Refactored RawMediaFrame to utilize ReadOnlySequence, and introduced PooledBufferWriter and PooledSequence for efficient memory management. Updated various payload classes to leverage new memory management strategies, and optimized RTSPHttpTransport for data sending. Replaced IEnumerable<ReadOnlyMemory<byte>> with ReadOnlySequence<byte> in RTSPEventArgs for improved performance.

ngraziano commented:
Hi, 

Thanks for the pull request but I cannot merge it.

 - The class PooledByteBuffer is not well tested, there is a problem because ArrayPool.Rent can return an array greater than the requested size and you return the all buffer in GetReadOnlySequence except for the last one.
 
 - This PR change the PR change the behavior of ProcessPacket, for example in H264, the data of the return RawMediaFrame is a list of NAL, each one is independent, your modification do not allow to find each NAL (may be important for SPS and PPS that are sent in the same RTP packet)

paulomorgado commented:
Hi,

> * This PR change the PR change the behavior of ProcessPacket, for example in H264, the data of the return RawMediaFrame is a list of NAL, each one is independent, your modification do not allow to find each NAL (may be important for SPS and PPS that are sent in the same RTP packet)

Is it important to index to each packet or is it OK to just enumerate?

ngraziano commented:
> Is it important to index to each packet or is it OK to just enumerate?

It's ok to just enumerate

paulomorgado commented:
Hi @ngraziano,

- Changed the `ReadOnlySequence<byte>` buffer to maintain the segments as they are added.
- Replaced `MemoryStream` with a segmented pooled `IBufferWriter<byte>`.

Can you check it out?

ngraziano commented:
I will take a look but it will take some time the commit is quite big

paulomorgado commented:
Hi  @ngraziano, have you had any time to look into this?

ngraziano commented:
Hi, I have look your PR, I'm not convince by the fact to use ReadOnlySequence and there is some data copy added of the video stream added. 
I will not merge it.

paulomorgado commented:
> Hi, I have look your PR, I'm not convince by the fact to use ReadOnlySequence and there is some data copy added of the video stream added. I will not merge it.

Where is the data copy added?

ngraziano commented:
> Where is the data copy added?
at `var data = nalUnitsBuffer.Clone();` in H264Payload

paulomorgado commented:
> > Where is the data copy added?
> > at `var data = nalUnitsBuffer.Clone();` in H264Payload

Good catch.

That could easily be fixed, but I don't think that's the issue here.
